File: throwcode_python/algorithm/sort/merge_sort.py
#!/use/bin/env/python3
# -*- coding:utf-8 -*-
# @author: goinhn
# @date: 2021/2/27 0:17

def merge_sort(arr):
    """
    归并排序
    :param arr:
    :return:
    """
    if arr is None or len(arr) < 2:
        return
    merge_sort_recursion(arr, 0, len(arr) - 1)


def merge_sort_recursion(arr, l, r):
    if l == r:
        return

    m = (l + r) // 2
    merge_sort_recursion(arr, l, m)
    merge_sort_recursion(arr, m + 1, r)
    merge(arr, l, m, r)


def merge(arr, l, m, r):
    arr_temp = [None] * (r - l + 1)
    arr_temp_index = 0
    l_temp = l
    r_temp = m + 1
    while l_temp < m + 1 and r_temp < r + 1:
        if arr[l_temp] < arr[r_temp]:
            arr_temp[arr_temp_index] = arr[l_temp]
            l_temp = l_temp + 1
        else:
            arr_temp[arr_temp_index] = arr[r_temp]
            r_temp = r_temp + 1
        arr_temp_index = arr_temp_index + 1

    while l_temp < m + 1:
        arr_temp[arr_temp_index] = arr[l_temp]
        l_temp = l_temp + 1
        arr_temp_index = arr_temp_index + 1

    while r_temp < r + 1:
        arr_temp[arr_temp_index] = arr[r_temp]
        r_temp = r_temp + 1
        arr_temp_index = arr_temp_index + 1

    for i in range(len(arr_temp)):
        arr[l + i] = arr_temp[i]


def main():
    arr = [2, 3, 1, 5, 4]
    merge_sort(arr)
    for value in arr:
        print(value)


if __name__ == '__main__':
    main()

File: throwcode_java/src/main/java/com/goinhn/algorithm/sort/MergeSort.java
package com.goinhn.algorithm.sort;

/**
 * <p>
 * 归并排序
 * 递归思想
 * </p>
 * <p>
 * 时间复杂度 O(N*logN)
 * 空间复杂度 O(N)
 * </p>
 * <p>
 * 注意数组转换的时候下标的位置
 * </p>
 *
 * @author goinhn
 * @date 2020-08-23T10:53:57
 */
public class MergeSort {

    public static void main(String[] args) {
        int[] arr = new int[]{1, 2, 3, 1, 6, 3};
        mergeSort(arr);
        for (int i : arr) {
            System.out.println(i);
        }
    }

    /**
     * 合并排序
     *
     * @param arr
     */
    public static void mergeSort(int[] arr) {
        if (arr == null || arr.length < 2) {
            return;
        }

        mergeSort(arr, 0, arr.length - 1);
    }

    /**
     * 递归方法
     *
     * @param arr
     * @param l
     * @param r
     */
    public static void mergeSort(int[] arr, int l, int r) {
        if (l == r) {
            return;
        }

        // 求平均，移位符号操作要加上括号()
        int m = l + ((r - l) >> 1);
        mergeSort(arr, l, m);
        mergeSort(arr, m + 1, r);
        merge(arr, l, m, r);
    }

    /**
     * 合并数组方法
     *
     * @param l
     * @param r
     * @return
     */
    public static void merge(int[] arr, int l, int m, int r) {
        int[] arrTemp = new int[r - l + 1];
        // 临时存储数组指针
        int index = 0;
        // 左数组的指针
        int pl = l;
        // 右数组的指针
        int pr = m + 1;

        while (pl <= m && pr <= r) {
            arrTemp[index++] = arr[pl] < arr[pr] ? arr[pl++] : arr[pr++];
        }

        while (pl <= m) {
            arrTemp[index++] = arr[pl++];
        }

        while (pr <= r) {
            arrTemp[index++] = arr[pr++];
        }

        for (int i = 0; i < arrTemp.length; i++) {
            arr[l + i] = arrTemp[i];
        }
    }

}
